### PR TITLE
Timeout modal

### PR DESCRIPTION
**Pull request by boxadesign**
### What is the context of this PR?
A timeout warning that lets users extend/reset the session.

- Two new components - generic modal (hidden from DS website) and timeout modal.
- Pattern - Extend session

### How to review
Read the documentation to see functionality, configuration and accessibility for testing at /components/timeout-modal/

### CODE REVIEW

**Review by jrbarnes9 — CHANGES_REQUESTED** [review]

File: src/patterns/help-users-to/extend-session/index.njk [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+### Focusing the page
+If the user leaves the tab or window, or their device goes to sleep, the page will lose focus. The session expiry time will be reset if the page is put back in focus before the session has expired.
+
+### Closing the timeout warning
+
+If the page has not been interacted with, the timeout warning will be displayed for a set number of seconds before the session expiry time. The user can reset the session time by closing the warning by selecting the primary action (“Continue” button) or by pressing the keyboard’s “Esc” key.
+
+## How to use this pattern
+
+Use the [timeout modal component](/components/timeout-modal) for the warning by importing it into the template you are using for any pages with a session expiry. For example, question pages. You should set the modal to be displayed for at least 60 seconds.
```
Comment (jrbarnes9) [inline src/patterns/help-users-to/extend-session/index.njk:50]:
```suggestion
Use the [timeout modal component](/components/timeout-modal) for the warning by importing it into the template you are using for any pages with a session expiry. You should set the modal to display for at least 60 seconds.
```

File: src/components/timeout-modal/index.njk [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+A timeout warning that lets users extend the amount of time they have to complete a task.
+
+{{
+    patternlibExample({ "path": "components/timeout-modal/examples/timeout/index.njk", "height": 500 })
+}}
+
+## When to use this component
+
+Use this component when a service needs to set a session expiry time to protect the users information.
```
Comment (jrbarnes9) [inline src/components/timeout-modal/index.njk:21]:
```suggestion
Use this component when a service sets a session expiry time to protect the user’s information.
```

File: src/components/timeout-modal/index.njk [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+## How to use this component
+
+The content of the timeout modal is set using the parameters shown in the macro options table in the previous example.
+
+You will also need to set the parameters:
+
+- The `showModalTimeInSeconds` with the number of seconds the modal should show before the session expiry time (this should be at least 60 seconds)
+- The `serverSessionExpiryEndpoint` with a relative url to the server’s endpoint for the component to send requests to
+- The `redirectUrl` with a url to the [session timed out error page](/error-status-pages/#session-timed-out-error-when-user-is-not-signed-in) the user is directed to when the session has timed out
```
Comment (jrbarnes9) [inline src/components/timeout-modal/index.njk:33]:
```suggestion
- `showModalTimeInSeconds` with the number of seconds the modal should show for before the session expiry time (this should be at least 60 seconds)
- `serverSessionExpiryEndpoint` with a relative url to the server’s API endpoint for the component to send requests to
- `redirectUrl` with a url to the [session timed out error page](/error-status-pages/#session-timed-out-error-when-user-is-not-signed-in) the user is directed to when the session has timed out
```

File: src/components/timeout-modal/index.njk [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+### Server requirements
+
+The timeout modal component requires an endpoint that handles two request methods and returns a JSON object containing the key `expires_at` with a value for the session expiry time in ISO format.
```
Comment (jrbarnes9) [inline src/components/timeout-modal/index.njk:37]:
```suggestion
The timeout modal component requires an API endpoint that handles two request methods and returns a JSON object containing the key `expires_at` with a value for the session expiry time in ISO format.
```

File: src/components/timeout-modal/index.njk [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+- the page receives focus, for example, by returning from another tab or when the device wakes
+- the user interacts with the page in any way, for example, by moving the cursor
+
+To efficiently handle the number of requests made when the user interacts with the page, the requests are "throttled" so only one is sent every minute, and others in the queue are cleared.
+
+#### 2. A request for the current expiry time
+
+When a request using a `GET` method is received, the response should contain the current expiry time.
+
+The request is made to check if the same session is being interacted within another tab. If the expiry time has changed, the timeout will be restarted against the expiry time in the response.
```
Comment (jrbarnes9) [inline src/components/timeout-modal/index.njk:56]:
```suggestion
The request is made to check if the same session is being interacted with in another tab. If the expiry time has changed, the timeout will be restarted against the expiry time in the response.
```

File: src/components/timeout-modal/index.njk [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+A request to get the current expiry time will be sent when the following happens:
+
+- before the modal opens
+- at 20 second intervals while the modal is open
```
Comment (jrbarnes9) [inline src/components/timeout-modal/index.njk:61]:
```suggestion
- at 20-second intervals while the modal is open
```

File: src/components/timeout-modal/index.njk [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+- before the page starts to redirect
+
+#### Failure responses from the server
+
+If the response status does not have a successful `200` code, the module component will handle the response as a session expiry and redirect the user to the url set in the `redirectUrl` parameter.
```
Comment (jrbarnes9) [inline src/components/timeout-modal/index.njk:66]:
```suggestion
If the response status does not have a successful `200` code, the component will handle the response as a session expiry and redirect the user to the url set in the `redirectUrl` parameter.
```

File: src/components/timeout-modal/index.njk [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        "linkText": "GDS timeout modal prototype and research"
+    })
+}}. We will be carrying out our own accessibility testing to validate it.
+
+
+### Accessibility requirements
+
+The WCAG  guidelines state:
+
+- There should be no limit to the number of times users can extend their session
```
Comment (jrbarnes9) [inline src/components/timeout-modal/index.njk:84]:
```suggestion
- there should be no limit to the number of times a user can extend their session
```

File: src/components/timeout-modal/index.njk [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+- The timeout modal should announce the remaining time at appropriate intervals to assistive technology
+- Make the underlying page content inert and focus the available action on the modal e.g. the "Continue" button
+- Focus on the previously focused element on the page when the modal is closed
```
Comment (jrbarnes9) [inline src/components/timeout-modal/index.njk:87]:
```suggestion
- the timeout modal should announce the remaining time at appropriate intervals to assistive technology
- the underlying page content should be made inert, and the available action on the modal must be put in focus, for example, the “Continue” button
- when the modal is closed the previously focused element on the page must be put back in focus
```

File: src/components/timeout-modal/index.njk [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+
+#### List of accessibility attributes on the component
+
+| Element                | Attribute and value                         | Description                                                                                           |
+| ---------------------- | ------------------------------------------- | ----------------------------------------------------------------------------------------------------- |
+| `<dialog>`             | `aria-labelledby=”ons-modal-title”`         | To make the contained heading be read out as title of the modal                                       |
```
Comment (jrbarnes9) [inline src/components/timeout-modal/index.njk:94]:
```suggestion
| `<dialog>`             | `aria-labelledby=”ons-modal-title”`         | To make the contained heading be read out as the title of the modal                                       |
```

File: src/components/timeout-modal/index.njk [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+| `<dialog>`             | `role="dialog"`	                           | Makes sure the modal is backwards compatible if the screen reader does not support the html element   |
+| `<dialog>`             | `aria-describedby="timeout-time-remaining"` | Reads out the live timer at 15 second intervals                                                       |
```
Comment (jrbarnes9) [inline src/components/timeout-modal/index.njk:96]:
```suggestion
| `<dialog>`             | `aria-describedby="timeout-time-remaining"` | Reads out the live timer at 15-second intervals                                                       |
```

File: src/components/timeout-modal/index.njk [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+| Countdown timer        | `aria-hidden="true"`	                       | Visible countdown timer is hidden from screen readers because it would be too interruptive            |
+| Countdown timer        | `aria-relevant=”additions”`	               | A workaround for aria-hidden for unsupported JAWS screen reader versions 17 and 18                    |
```
Comment (jrbarnes9) [inline src/components/timeout-modal/index.njk:98]:
```suggestion
| Countdown timer        | `aria-relevant=”additions”`	               | A workaround for `aria-hidden` for unsupported JAWS screen reader versions 17 and 18                    |
```

**Review by jrbarnes9 — CHANGES_REQUESTED** [review]

File: src/patterns/help-users-to/extend-session/examples/timeout/index.njk [hunk 13]
```diff
@@ -27,9 +25,12 @@ layout: none
         "title": "On 1 May 2016, what was the number of employees for Bolt and Ratchet?",
         "description": "<p>This is all employees aged 16 years or over that your organisation employs.</p>"
```
Comment (jrbarnes9) [inline src/patterns/help-users-to/extend-session/examples/timeout/index.njk:26]: Do you need `"legendIsQuestionTitle": true` here?

**Review by jrbarnes9 — DISMISSED** [review]

**Review by jrbarnes9 — APPROVED** [review]

File: src/tests/spec/autosuggest/autosuggest.ui.specdisabled.js [hunk 14]
```diff
@@ -527,7 +527,7 @@ describe('Autosuggest.ui component', function() {
         });
 
         it('then the function should resolve', function() {
-          this.autosuggest.fetchSuggestions('yes').should.eventually.eql(this.result);
+          this.autosuggest.fetchSuggestions('yes').should.y.eql(this.result);
```
Comment (rmccar) [inline src/tests/spec/autosuggest/autosuggest.ui.specdisabled.js:530]: Is this a typo?

File: src/components/timeout-modal/index.njk [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+
+## How to use this component
+
+The content of the timeout modal is set using the parameters shown in the macro options table in the previous example.
+
+You will also need to set the parameters:
+
+- `showModalTimeInSeconds` with the number of seconds the modal should show for before the session expiry time (this should be at least 60 seconds)
+- `serverSessionExpiryEndpoint` with a relative url to the server’s API endpoint for the component to send requests to
+- `redirectUrl` with a url to the [session timed out error page](/error-status-pages/#session-timed-out-error-when-user-is-not-signed-in) the user is directed to when the session has timed out
```
Comment (rmccar) [inline src/components/timeout-modal/index.njk:33]: session timed out error page link is broken

File: src/tests/spec/autosuggest/autosuggest.ui.specdisabled.js (hunk 14 above)
Comment (boxadesign) [reply src/tests/spec/autosuggest/autosuggest.ui.specdisabled.js:530]: Yes, those tests are currently disabled at the moment.

**Review by rmccar — APPROVED** [review]

**Pull request merged**